### ISSUE TITLE
Cut wing at given eta position

### ISSUE THREAD

**Issue by MarAlder** [DLR-SL/CPACS-IFARX#6]
In the webinar Vladimir asked the question of how to get the profile at a given eta position if no section is given there. 

I added an example to [examles/cut_wing](https://github.com/DLR-SL/CPACS-IFARX/tree/master/examples/cut_wing) which shows how to use openCascade to cut the wing and discretize the resulting face.

![wing_cut](https://user-images.githubusercontent.com/43143741/66412038-0c0dc880-e9f5-11e9-9a32-fa9e2981eb0d.png)

**Comment by godaygo** [comment]
Marko thanks for example on how to use _openCascades_. While your solution feels rather general and and maybe it is really the only right way to proceed, it is also requires quite good familiarity with _python-occ_. I was able to do the same just by using only `tiglwrapper` but maybe the way I'm doing this is error prone or wrong? So it would be nice to know about possible shortcomings. Below is my version of extracting a wing\`s airfoil by specifying `eta`.

```python
from tixi import tixiwrapper
from tigl import tiglwrapper

import bisect
import numpy as np
import matplotlib.pyplot as plt

tixi = tixiwrapper.Tixi()
tixi.open('./data/aircraft.xml')
tigl = tiglwrapper.Tigl()
tigl.open(tixi, 'AircraftModel')

wing_uid = 'wing'

def get_airfoil(wing_uid, eta):
    assert 0.0 <= eta <= 1.0
    windex = tigl.wingGetIndex(wing_uid)
    nseg = tigl.wingGetSegmentCount(windex)
    bounds = [tigl.wingGetChordPoint(windex, idx, 0., 0.)[1] for idx in range(1, nseg + 1)]
    bounds += [tigl.wingGetChordPoint(windex, nseg, 1., 0.)[1]]
    span_pos = eta*bounds[-1]
    seg_idx = bisect.bisect(bounds, span_pos)
    seg_len = bounds[seg_idx] - bounds[seg_idx - 1]
    loc_eta = (span_pos - bounds[seg_idx - 1])/seg_len
    upper_pts  = []
    lower_pts = []
    for xsi in np.linspace(0., 1., 51):
        upper_pts.append(tigl.wingGetUpperPoint(windex, seg_idx, loc_eta, xsi))
        lower_pts.append(tigl.wingGetLowerPoint(windex, seg_idx, loc_eta, xsi))
    airfoil = np.vstack([upper_pts, lower_pts])
    return airfoil

# get_airfoil('wing', 0.5)
```
While doing this I've also found some strange things:

1)
I don't know the reasons why these results differ, I expected that they should give the same results, but numerically the difference is pretty huge.
```python
>>> 0.5*eta*tigl.wingGetSpan('wing')
17.991717736644723
>>> tigl.wingGetChordPoint(3, 4, 1., 0.)[1]
17.9910459436
```
At the same time in CPACS file (`aircraft.xml`) the translation of `<section uID="wing_Sec5">` section is:
```xml
<translation refType="absGlobal">
     <x>8.26651530788</x>
     <y>17.9910459436</y>
     <z>1.93606028828</z>
 ``` 
The value is the same as with `wingGetChordPoint`. So there is something strange with `wingGetSpan`, also please look at point 3 below.

2)
Functions in `tiglwrapper`  that use `eta` and `xsi`, will throw an error if you provide a value outside the range of `[0, 1]`. But you can specify any `eta` and `xsi` for `tigl.wingGetChordPoint` which was rather unexpected.

3)
Almost all functions use `index` to refer to part (wing, htp, ...) or segment. But for some reasons `tigl.wingGetSpan` uses `wingUID`.

**Comment by godaygo** [comment]
@MarAlder I've played with your code a bit more, and found bug in my snippet :) So the updated version of function is:
```python
def get_airfoil(wing_uid, eta):
    assert 0.0 <= eta <= 1.0
    windex = tigl.wingGetIndex(wing_uid)
    nseg = tigl.wingGetSegmentCount(windex)
    bounds = [tigl.wingGetChordPoint(windex, idx, 0., 0.)[1] for idx in range(1, nseg + 1)]
    bounds += [tigl.wingGetChordPoint(windex, nseg, 1., 0.)[1]]
    span_pos = eta*bounds[-1]
    seg_idx = bisect.bisect(bounds, span_pos) if eta < 1.0 else nseg
    seg_len = bounds[seg_idx] - bounds[seg_idx - 1]
    loc_eta = (span_pos - bounds[seg_idx - 1])/seg_len
    upper_pts  = []
    lower_pts = []
    for xsi in np.linspace(0., 1., 101):
        lower_pts.append(tigl.wingGetLowerPoint(windex, seg_idx, loc_eta, xsi))
        upper_pts.append(tigl.wingGetUpperPoint(windex, seg_idx, loc_eta, 1 - xsi))
    airfoil = np.vstack([lower_pts, upper_pts])
    return airfoil
```
I've also changed the points order to be CCW like in your example. Comparing the results of these both solutions I basically made sure that they do the same thing. The difference in `y` coordinate is because of that fact:
```python
...
>>> wing_section = slicing(wing_shape, 0.99)
>>> pnts, face_ids = discretize(wing_section, 1e-5) 
>>> my_pts = get_airfoil('wing', 0.99)
>>> pnts[0, 1] / my_pts[0, 1]
1.0000373404512357
>>> 0.5*tigl.wingGetSpan('wing') / tigl.wingGetChordPoint(3, 4, 1., 0.)[1]
1.0000373404107148
...
```
The is because what _semispan is the right one_, of course I would like to which. 

p.s.: my previous version fails with `eta` close to `1.` now it is fixed. Currently yours produces empty array for `eta` close to `1.` :wink:
```python
>>> wing_section = slicing(wing_shape, 0.99998)
>>> pnts, face_ids = discretize(wing_section, 1e-9)
>>> pnts
array([], shape=(0, 3), dtype=float64)
```

**Comment by MarAlder** [comment]
Hi Kirill,

I like your code much more than my own one :smile:. My example was rather using openCascade, because it's an extract from a code were I needed intersections with more general shapes. You're making the TiGL developers happy by using the more convenient TiGL functions.  

Differences between the approaches originate from the following: [`tigl.wingGetSpan`](https://github.com/DLR-SC/tigl/blob/76e1f1442a045e1b8b7954119ca6f9c883ea41e2/src/wing/CCPACSWing.cpp#L475-L558) places a OCC bounding box around the wing. [`GetChordPoint`](https://github.com/DLR-SC/tigl/blob/76e1f1442a045e1b8b7954119ca6f9c883ea41e2/src/wing/CCPACSWingSegment.cpp#L886-L892) is based on a [translation](https://github.com/DLR-SC/tigl/blob/76e1f1442a045e1b8b7954119ca6f9c883ea41e2/src/math/CTiglPointTranslator.cpp#L141-L195) of face coordinates. I believe this is the reason that normalized values below `0.` and above `1.` are not returning errors here. (I hope the explanation is reasonably correct? --> @rainman110, @joergbrech )

The OCC boundingBox algorithm does (numerically) not exactly return the y-position of the last segment This is the difference you observed. Maybe we should rethink the `getSpan` approach to get consistent values. However, this fidelity was not a problem for our OAD group yet. 

My code also uses the boundingBox. That's why the values slightly differ and `1.` returns an empty array in my code. I'll put your code in the examples folder if that's ok for you.

**Comment by MarAlder** [comment]
> Almost all functions use index to refer to part (wing, htp, ...) or segment. But for some reasons tigl.wingGetSpan uses wingUID.

Yep, we should maybe provide both :thinking:. I'm not sure why it differs from the other library functions.

**Comment by godaygo** [comment]
> I'll put your code in the examples folder if that's ok for you.

I'm totally ok with that. I'm thinking about some small modifications to make this function more usable. Like make it possible to specify a `xsi` array, which will allow to specify the concentration of points on the leading edge. In the current version while the order is CCW but `1-xsi` is used for upper side - it will not work. But I will have the opportunity to do this only tonight.

**Comment by MarAlder** [comment]
Ok, cool. I gave you write permissions, so just put it in the example folder when you are happy with it.

**Comment by MarAlder** [comment]
@godaygo I can recommend you to look through these [slides & examples](https://github.com/rainman110/tigl-workshop) if you are more interested in TiGL developments.

**Comment by godaygo** [comment]
Vladimir noticed that my code doesn't work for 'htp':

> Have u tried to run it for vtp?
> Did't work for me
> In the case of wing and htp works well. Great job

Yeah, that's true. And the problem was in this line because index `1` was hard-coded. https://github.com/DLR-SL/CPACS-IFARX/blob/e01a91ca597933638287cdb903efca3404b824bd/examples/cut_wing_2/get_airfoil_points.py#L79-L80
 I've just updated this part, so now everything works. But I must admit that I don't like the way I did it. To do this correctly, the introduction of a local coordinate system (LCS) is required, but this will complicate the implementation, and will distract from the main purpose of the example.

@MarAlder As far as I understand, at the moment, in CPACS there is no such thing as a local coordinate system, only local parameters (I wonder how they are defined internally)? In order to calculate LCS, I can just read the data from `<transformation>` fields, but maybe there is a better way to do it?
